### PR TITLE
Fixed template assignments with old basic and full names with tpl extension

### PR DESCRIPTION
**Pull request by MSeal**
In the refactor that simplified the template_name fetching we broke using `template=basic` despite the path to the compatibility templates being added to the search paths. This fixes this by searching the compatibility directory last.

Second half of resolving https://github.com/jupyter/nbconvert/issues/1384

### CODE REVIEW

File: nbconvert/exporters/templateexporter.py
Comment (MSeal) [inline nbconvert/exporters/templateexporter.py:567]: A minor optimization, this does change the algorithm to find the first instead of the last match. I _think_ this is fine but I'd like other eyes on it in case I need to reverse the iteration ordering.

**Comment by SylvainCorlay** [comment]
thanks for the ping. i will look into this Monday morning paris time.

**Review by willingc — APPROVED** [review]
LGTM @MSeal. Will leave for @SylvainCorlay to review and comment/merge.

**Comment by MSeal** [comment]
Hmm before merging the comment thread on https://github.com/jupyter/nbconvert/pull/1386 should be read. I think we (I?) eliminated the `base` template pattern and erroneously stated classic maps to that? I posted what I think works around it to get back to original functionality but I'd like some input and what we should do (because it impacts this backwards compatibility PR)

File: nbconvert/exporters/templateexporter.py
Comment (SylvainCorlay) [reply nbconvert/exporters/templateexporter.py:567]: I am not 100% sure about the `break` statement.
What if configuration is overriden in a higher precedence extra template dir?

Ping @maartenbreddels. Let me know what you think.

File: nbconvert/exporters/templateexporter.py
Comment (SylvainCorlay) [inline nbconvert/exporters/templateexporter.py:577]: Same here. Not sure the recursion on root_dirs should be conditional on `found_at_least_one`.

File: nbconvert/exporters/templateexporter.py
Comment (SylvainCorlay) [inline nbconvert/exporters/templateexporter.py:593]: This approach may collide with an actual template called `full` or `basic`.

Maybe we should condition on this at a higher level...

Comment (MSeal) [reply nbconvert/exporters/templateexporter.py:593]: I was thinking of changing the include paths inside template files to include the template name. I think that would resolve a few issues and not require these hacks for the backwards compatibility, as template extensions are similarly running into 2nd hop extension path failures.

Comment (MSeal) [reply nbconvert/exporters/templateexporter.py:593]: Updated PR with the change I discussed -- extending existing templates from custom templates no longer explodes when files are included across template boundaries. e.g. Fixes https://github.com/jupyter/nbconvert/issues/1394

**Comment by MSeal** [comment]
I added the basic template back as well, since it has a need and we made using it more difficult and broke backwards compatibility.

File: nbconvert/exporters/templateexporter.py
Comment (MSeal) [reply nbconvert/exporters/templateexporter.py:582]: The recursion isn't conditional, it recurses to completion before returning back to this context

File: share/jupyter/nbconvert/templates/classic/base.html.j2
Comment (SylvainCorlay) [inline share/jupyter/nbconvert/templates/classic/base.html.j2:2]: Why is it necessary to refer to base here?

**Comment by SylvainCorlay** [comment]
@MSeal it should not be required to add the path to the base template in jinja templates, unless you override them locally.

Does your change make it necessary?

File: share/jupyter/nbconvert/templates/classic/base.html.j2
Comment (SylvainCorlay) [reply share/jupyter/nbconvert/templates/classic/base.html.j2:2]: Actually, if these changes in the jinja templates are required, this means this is really backward incompatible.

**Comment by SylvainCorlay** [comment]
@maartenbreddels @MSeal would you like to have a synchronous discussion about how to fix this the best early next week?

File: nbconvert/exporters/templateexporter.py
Comment (SylvainCorlay) [reply nbconvert/exporters/templateexporter.py:572]: But it will not merge the configuration with other matches?

File: nbconvert/exporters/templateexporter.py
Comment (SylvainCorlay) [reply nbconvert/exporters/templateexporter.py:593]: > changing the include paths inside template files to include the template name

I am not sure about this, because it is a backward incompatible change with respect to the current behavior.

Comment (MSeal) [reply nbconvert/exporters/templateexporter.py:593]: How is it backwards incompatible? The templates behave identically within nbconvert but now when extended from external templates will behave correctly (since they don't have the proper template paths set to include this template). It also protects from multiple inheritance of file extensions -- e.g. there's ambiguity when importing null.js is some places if you didn't include the template name in the extension path.

File: nbconvert/exporters/templateexporter.py
Comment (MSeal) [reply nbconvert/exporters/templateexporter.py:572]: It wasn't before because it overwrote the `conf` and `conf_file` whenever it found a template until the last one found. I don't think it changes the behavior beyond preferring the first exact match over the last exact match. If you're really worried I can remove the optimization.

**Comment by MSeal** [comment]
> @MSeal it should not be required to add the path to the base template in jinja templates, unless you override them locally.
>
> Does your change make it necessary?

Only in the sense that the backwards compatibility templates were not working without it. Similarly there's several open issues pointing out that extending `lab/index.html.js` and other files within the templating system fail if those templates use relative extension paths within nbconvert.

> @maartenbreddels @MSeal would you like to have a synchronous discussion about how to fix this the best early next week?

Would have to be tomorrow (Sunday) or Tuesday morning as my Monday has no wiggle room on my end.

File: nbconvert/exporters/templateexporter.py
Comment (SylvainCorlay) [reply nbconvert/exporters/templateexporter.py:593]: It is backward incompatible for custom template authors who inherit from base.

**Comment by maartenbreddels** [comment]
Tuesday works for me.

I think we should not prefix with `/lab` `/base` in the extends, as we'll lose the ability to override those files from other directories.

Going over the code, and trying to make the test pass without the prefixes, I don't think we can go with full 5.x compatibility as it is now. I think if we want compatibility with full.tpl and base.tpl, we should just copy the old content of nbconvert 5.x, and keep/add the deprecation warning.

**Comment by MSeal** [comment]
> Tuesday works for me.

Invite sent.

> we'll lose the ability to override those files from other directories.

I see. Unfortunately since the template_paths don't dynamically reflect extensions made, supporting that pattern breaks the ability for folks to extend the `<template>/index.<format>.j2` files as the local extensions fail (not just for backwards compatibility). I didn't see a clean way to automatically add templates that were extended to the template_paths to prevent this issue.

**Comment by MSeal** [comment]
Closing as alternative fixes were merged

**Pull request closed without merging**
